### PR TITLE
Implement Guard Patterns down to THIR

### PR DESCRIPTION
**Pull request by max-niederman**
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/129996)*

Partial implementation of RFC 3637 including parsing, HIR lowering, and THIR lowering but not MIR lowering or anything depending on it.

See [the tracking issue](https://github.com/rust-lang/rust/issues/129967) for more info.

### CODE REVIEW

**Comment by max-niederman** [comment]
I unfortunately don't really have time to work on this anymore, so I'll be closing this at least for the foreseeable future.

File: compiler/rustc_mir_build/src/thir/cx/pattern/mod.rs [hunk 1]
```diff
@@ -407,9 +400,10 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
 
             hir::PatKind::Or(pats) => PatKind::Or { pats: self.lower_patterns(pats) },
 
-            hir::PatKind::Guard(..) => {
-                span_bug!(pat.span, "lowering guard patterns to THIR is not yet implemented");
-            }
+            hir::PatKind::Guard(subpat, condition) => PatKind::Guard {
+                subpattern: self.lower_pattern(subpat),
+                condition: self.cx.mirror_expr(condition),
```
Comment (Human9000-bit) [inline compiler/rustc_mir_build/src/thir/cx/pattern/mod.rs]: Was all this refactor needed because you couldn't call `mirror_expr()` otherwise?

**Pull request closed without merging**